File: server/src/rag.ts
import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";
import Database from "better-sqlite3";
import * as personality from "./personality.js";

const RAG_DIR = process.env.AI_PERSONALITY_DIR
  ? path.resolve(process.env.AI_PERSONALITY_DIR, "..", "rag")
  : path.join(os.homedir(), ".ai-personality", "rag");

const DB_PATH = path.join(RAG_DIR, "vector.db");

interface ChunkRow {
  id: number;
  filename: string;
  heading: string;
  content: string;
  embedding: Buffer;
  type: string;
}

export interface SearchResult {
  filename: string;
  heading: string;
  content: string;
  type: string;
  score: number;
}

let _pipe: any = null;
let _db: Database.Database | null = null;
let _watcher: fs.FSWatcher | null = null;

async function getPipe() {
  if (!_pipe) {
    const { pipeline: p } = await import("@xenova/transformers");
    _pipe = await p("feature-extraction", "Xenova/all-MiniLM-L6-v2");
  }
  return _pipe;
}

function getDb(): Database.Database {
  if (_db) return _db;
  if (!fs.existsSync(RAG_DIR)) {
    fs.mkdirSync(RAG_DIR, { recursive: true });
  }
  _db = new Database(DB_PATH);
  _db.exec(`CREATE TABLE IF NOT EXISTS chunks (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    filename TEXT NOT NULL,
    heading TEXT,
    content TEXT NOT NULL,
    embedding BLOB NOT NULL,
    type TEXT NOT NULL
  )`);
  return _db;
}

function chunkContent(filename: string, type: string, body: string): Array<{ filename: string; heading: string; content: string; type: string }> {
  const lines = body.split("\n");
  const chunks: Array<{ filename: string; heading: string; content: string; type: string }> = [];
  let currentHeading = "";
  let currentLines: string[] = [];

  for (const line of lines) {
    const headingMatch = line.match(/^(#{2,3})\s+(.+)/);
    if (headingMatch) {
      if (currentLines.length > 0) {
        chunks.push({ filename, heading: currentHeading, content: currentLines.join("\n").trim(), type });
      }
      currentHeading = headingMatch[2].trim();
      currentLines = [line];
    } else {
      currentLines.push(line);
    }
  }
  if (currentLines.length > 0) {
    chunks.push({ filename, heading: currentHeading, content: currentLines.join("\n").trim(), type });
  }
  return chunks;
}

export async function embed(text: string): Promise<Float32Array> {
  const pipe = await getPipe();
  const result = await pipe(text, { pooling: "mean", normalize: true });
  return result.data as Float32Array;
}

export async function buildIndex(): Promise<number> {
  const db = getDb();
  db.exec("DELETE FROM chunks");

  const files = personality.readAllPersonalityFiles();
  const insert = db.prepare("INSERT INTO chunks (filename, heading, content, embedding, type) VALUES (?, ?, ?, ?, ?)");

  let total = 0;
  for (const f of files) {
    const chunks = chunkContent(f.filename, f.frontmatter.type, f.body);
    for (const chunk of chunks) {
      try {
        const emb = await embed(chunk.content.slice(0, 512));
        const buffer = Buffer.from(emb.buffer);
        insert.run(chunk.filename, chunk.heading, chunk.content, buffer, chunk.type);
        total++;
      } catch (err) {
        console.error(`Failed to embed chunk from ${f.filename}:`, err);
      }
    }
  }
  console.error(`Indexed ${total} chunks`);
  return total;
}

export async function search(query: string, topK: number = 5): Promise<SearchResult[]> {
  const db = getDb();
  const rows = db.prepare("SELECT * FROM chunks").all() as ChunkRow[];
  if (rows.length === 0) return [];

  const queryEmb = await embed(query);
  const scored = rows.map((row) => {
    const storedEmb = new Float32Array(row.embedding.buffer, row.embedding.byteOffset, row.embedding.byteLength / 4);
    return { row, score: cosineSimilarity(queryEmb, storedEmb) };
  });

  scored.sort((a, b) => b.score - a.score);
  return scored.slice(0, topK).map((r) => ({
    filename: r.row.filename,
    heading: r.row.heading,
    content: r.row.content,
    type: r.row.type,
    score: r.score,
  }));
}

export function searchWithEmbedding(queryEmb: Float32Array, topK: number = 5): SearchResult[] {
  const db = getDb();
  const rows = db.prepare("SELECT * FROM chunks").all() as ChunkRow[];
  if (rows.length === 0) return [];

  const scored = rows.map((row) => {
    const storedEmb = new Float32Array(row.embedding.buffer, row.embedding.byteOffset, row.embedding.byteLength / 4);
    return { row, score: cosineSimilarity(queryEmb, storedEmb) };
  });

  scored.sort((a, b) => b.score - a.score);
  return scored.slice(0, topK).map((r) => ({
    filename: r.row.filename,
    heading: r.row.heading,
    content: r.row.content,
    type: r.row.type,
    score: r.score,
  }));
}

function cosineSimilarity(a: Float32Array, b: Float32Array): number {
  let dot = 0, na = 0, nb = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    na += a[i] * a[i];
    nb += b[i] * b[i];
  }
  return dot / (Math.sqrt(na) * Math.sqrt(nb));
}

export function startWatcher(): void {
  if (_watcher) return;
  const dir = personality.PERSONALITY_DIR;
  if (!fs.existsSync(dir)) return;

  _watcher = fs.watch(dir, (eventType, filename) => {
    if (filename && filename.endsWith(".md") && eventType === "change") {
      console.error(`Personality file changed: ${filename}, reindexing...`);
      buildIndex().then((n) => console.error(`Reindexed ${n} chunks after change to ${filename}`));
    }
  });
}

export function stopWatcher(): void {
  if (_watcher) {
    _watcher.close();
    _watcher = null;
  }
}

export function close(): void {
  stopWatcher();
  if (_db) {
    _db.close();
    _db = null;
  }
}

File: server-py/src/ai_personality/rag.py
import os
import time
import sqlite3
import struct
import threading
from pathlib import Path

from . import personality

_RAG_DIR = None
_DB: sqlite3.Connection | None = None
_MODEL = None
_WATCHER_THREAD: threading.Thread | None = None
_WATCHER_STOP = threading.Event()
_WATCHER_MTIMES: dict[str, float] = {}


def _get_rag_dir() -> Path:
    global _RAG_DIR
    if _RAG_DIR is not None:
        return _RAG_DIR
    env = os.environ.get("AI_PERSONALITY_DIR")
    if env:
        _RAG_DIR = Path(env).resolve().parent / "rag"
    else:
        _RAG_DIR = Path.home() / ".ai-personality" / "rag"
    _RAG_DIR.mkdir(parents=True, exist_ok=True)
    return _RAG_DIR


def _get_db() -> sqlite3.Connection:
    global _DB
    if _DB is not None:
        return _DB
    db_path = _get_rag_dir() / "vector.db"
    _DB = sqlite3.connect(str(db_path))
    _DB.execute(
        "CREATE TABLE IF NOT EXISTS chunks ("
        "  id INTEGER PRIMARY KEY AUTOINCREMENT,"
        "  filename TEXT NOT NULL,"
        "  heading TEXT,"
        "  content TEXT NOT NULL,"
        "  embedding BLOB NOT NULL,"
        "  type TEXT NOT NULL"
        ")"
    )
    return _DB


def _get_model():
    global _MODEL
    if _MODEL is not None:
        return _MODEL
    from fastembed import TextEmbedding

    _MODEL = TextEmbedding("BAAI/bge-small-en-v1.5")
    return _MODEL


def _chunk_content(filename: str, type_name: str, body: str) -> list[dict]:
    lines = body.split("\n")
    chunks: list[dict] = []
    current_heading = ""
    current_lines: list[str] = []

    for line in lines:
        m = __import__("re").match(r"^(#{2,3})\s+(.+)", line)
        if m:
            if current_lines:
                chunks.append(
                    {
                        "filename": filename,
                        "heading": current_heading,
                        "content": "\n".join(current_lines).strip(),
                        "type": type_name,
                    }
                )
            current_heading = m.group(2).strip()
            current_lines = [line]
        else:
            current_lines.append(line)

    if current_lines:
        chunks.append(
            {
                "filename": filename,
                "heading": current_heading,
                "content": "\n".join(current_lines).strip(),
                "type": type_name,
            }
        )

    return chunks


def _float32_to_blob(vec: list[float]) -> bytes:
    return struct.pack(f"{len(vec)}f", *vec)


def _blob_to_float32(blob: bytes) -> list[float]:
    return list(struct.unpack(f"{len(blob) // 4}f", blob))


def _cosine_similarity(a: list[float], b: list[float]) -> float:
    dot = na = nb = 0.0
    for i in range(len(a)):
        dot += a[i] * b[i]
        na += a[i] * a[i]
        nb += b[i] * b[i]
    return dot / ((na ** 0.5) * (nb ** 0.5))


def build_index() -> int:
    db = _get_db()
    db.execute("DELETE FROM chunks")

    model = _get_model()
    files = personality.read_all_personality_files()
    total = 0

    for f in files:
        chunks = _chunk_content(f["filename"], f["frontmatter"].get("type", "?"), f["body"])
        texts = [c["content"][:512] for c in chunks]

        for chunk, embedding in zip(chunks, model.embed(texts)):
            blob = _float32_to_blob(list(embedding))
            db.execute(
                "INSERT INTO chunks (filename, heading, content, embedding, type) VALUES (?, ?, ?, ?, ?)",
                (chunk["filename"], chunk["heading"], chunk["content"], blob, chunk["type"]),
            )
            total += 1

    db.commit()
    print(f"Indexed {total} chunks", file=__import__("sys").stderr)
    return total


def search(query: str, top_k: int = 5) -> list[dict]:
    db = _get_db()
    rows = db.execute("SELECT * FROM chunks").fetchall()
    if not rows:
        return []

    model = _get_model()
    query_emb = list(next(model.embed([query])))
    columns = [desc[0] for desc in db.execute("SELECT * FROM chunks").description]

    scored: list[tuple[dict, float]] = []
    for row in rows:
        chunk = dict(zip(columns, row))
        stored = _blob_to_float32(chunk["embedding"])
        score = _cosine_similarity(query_emb, stored)
        scored.append((chunk, score))

    scored.sort(key=lambda x: x[1], reverse=True)
    return [
        {
            "filename": chunk["filename"],
            "heading": chunk["heading"],
            "content": chunk["content"],
            "type": chunk["type"],
            "score": score,
        }
        for chunk, score in scored[:top_k]
    ]


def _watcher_loop():
    dir_path = personality.PERSONALITY_DIR

    while not _WATCHER_STOP.is_set():
        if dir_path.exists():
            for fname in personality.FILE_NAMES:
                fp = dir_path / fname
                if fp.exists():
                    mtime = fp.stat().st_mtime
                    prev = _WATCHER_MTIMES.get(fname)
                    if prev is not None and mtime != prev:
                        print(f"Personality file changed: {fname}, reindexing...", file=__import__("sys").stderr)
                        try:
                            n = build_index()
                            print(f"Reindexed {n} chunks after change to {fname}", file=__import__("sys").stderr)
                        except Exception as e:
                            print(f"Reindex failed: {e}", file=__import__("sys").stderr)
                    _WATCHER_MTIMES[fname] = mtime
        _WATCHER_STOP.wait(2.0)


def start_watcher():
    global _WATCHER_THREAD
    if _WATCHER_THREAD is not None:
        return

    dir_path = personality.PERSONALITY_DIR
    if dir_path.exists():
        for fname in personality.FILE_NAMES:
            fp = dir_path / fname
            if fp.exists():
                _WATCHER_MTIMES[fname] = fp.stat().st_mtime

    _WATCHER_STOP.clear()
    _WATCHER_THREAD = threading.Thread(target=_watcher_loop, daemon=True)
    _WATCHER_THREAD.start()


def stop_watcher():
    _WATCHER_STOP.set()
    global _WATCHER_THREAD
    if _WATCHER_THREAD:
        _WATCHER_THREAD.join(timeout=3)
        _WATCHER_THREAD = None


def close():
    stop_watcher()
    global _DB
    if _DB:
        _DB.close()
        _DB = None
